shenyu-loadbalancer/src/main/java/org/apache/shenyu/loadbalancer/spi/RandomLoadBalancer.java

```diff
@@ -41,42 +42,41 @@ public Upstream doSelect(final List<Upstream> upstreamList, final String ip) {
         weights[0] = firstUpstreamWeight;
         // init the totalWeight
         int totalWeight = firstUpstreamWeight;
-        int halfLengthTotalWeight = 0;
         for (int i = 1; i < length; i++) {
             int currentUpstreamWeight = getWeight(upstreamList.get(i));
-            if (i <= (length + 1) / 2) {
-                halfLengthTotalWeight = totalWeight;
-            }
-            weights[i] = currentUpstreamWeight;
             totalWeight += currentUpstreamWeight;
+            weights[i] = totalWeight;
             if (sameWeight && currentUpstreamWeight != firstUpstreamWeight) {
                 // Calculate whether the weight of ownership is the same.
                 sameWeight = false;
             }
         }
         if (totalWeight > 0 && !sameWeight) {
-            return random(totalWeight, halfLengthTotalWeight, weights, upstreamList);
+            return random(totalWeight, weights, upstreamList, length);
         }
         return random(upstreamList);
     }
 
-    private Upstream random(final int totalWeight, final int halfLengthTotalWeight, final int[] weights, final List<Upstream> upstreamList) {
+    private Upstream random(final int totalWeight, final int[] weights, final List<Upstream> upstreamList, final int length) {
         // If the weights are not the same and the weights are greater than 0, then random by the total number of weights.
         int offset = RANDOM.nextInt(totalWeight);
-        int index = 0;
-        int end = weights.length;
-        if (offset >= halfLengthTotalWeight) {
-            index = (weights.length + 1) / 2;
-            offset -= halfLengthTotalWeight;
+        if (length <= 4) {
```

Comment:
why 4?

Reply:
Binary search has advantages only when the array length is long. The 4 here are just personal preferences

Reply:
when service lists are 5?

Reply:
> when service lists are 5?

Then it will be searched through binary search